Supervised Learning Model

In [1]:
import pandas as pd 
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split 
import matplotlib.pyplot as plt
%matplotlib inline 
from sklearn import metrics 
import numpy as np 

In [2]:
pokemon = pd.read_csv("../Data/300k.csv", low_memory=False)
pokemon

pokemonId   latitude   longitude    appearedLocalTime  \
0              16  20.525745  -97.460829  2016-09-08T03:57:45   
1             133  20.523695  -97.461167  2016-09-08T03:57:37   
2              16  38.903590  -77.199780  2016-09-08T03:57:25   
3              13  47.665903 -122.312561  2016-09-08T03:56:22   
4             133  47.666454 -122.311628  2016-09-08T03:56:08   
...           ...        ...         ...                  ...   
296016         19  39.119052  -76.633194  2016-09-02T21:50:11   
296017         13  47.142237    7.244752  2016-09-02T21:50:08   
296018         16  42.258532  -87.881248  2016-09-02T21:50:07   
296019         19  26.230929  -98.194772  2016-09-02T21:50:07   
296020         96  50.144272   14.101178  2016-09-02T21:50:06   

                                 _id    cellId_90m   cellId_180m  \
0       NTgxMDkzOTk4MTM5MjUwMjIzNw==  9.645140e+18  9.645140e+18   
1       OTQ1NDgzODc1MjM3NDEzMTI2MQ==  9.645140e+18  9.645140e+18   
2       NTQ0OTQ0NDA1Nzg2ODg3OTg2OQ==  9.923200e+18  9.923200e+18   
3       NTU2MTU1NDM4NzA2MDk1MDcxNw==  6.093390e+18  6.093390e+18   
4       MTY2ODg4MTAzMTczMDE0MTUwNTM=  6.093390e+18  6.093390e+18   
...                              ...           ...           ...   
296016  MzM1MzA0NjEyNTIxNTQ1OTY5Mw==  9.923680e+18  9.923680e+18   
296017  NjA0NDc3OTcxMzk2NjM5MzgyMQ==  5.156090e+18  5.156090e+18   
296018  OTIzODI4ODYyOTUyMjI4NzI0NQ==  9.804220e+18  9.804220e+18   
296019  MTcwMjEwNDQ3Mjk1MjIwOTM5NDk=  9.684320e+18  9.684320e+18   
296020  MTMzNzM3NzQ1MTM4MTcwMTE0Njk=  5.119390e+18  5.119390e+18   

         cellId_370m   cellId_730m  cellId_1460m  ...  cooc_143  cooc_144  \
0       9.645140e+18  9.645140e+18  9.645140e+18  ...     False     False   
1       9.645140e+18  9.645140e+18  9.645140e+18  ...     False     False   
2       9.923200e+18  9.923200e+18  9.923200e+18  ...     False     False   
3       6.093390e+18  6.093390e+18  6.093390e+18  ...     False     False   
4       6.093390e+18  6.093390e+18  6.093390e+18  ...     False     False   
...              ...           ...           ...  ...       ...       ...   
296016  9.923680e+18  9.923680e+18  9.923680e+18  ...     False     False   
296017  5.156090e+18  5.156090e+18  5.156090e+18  ...     False     False   
296018  9.804220e+18  9.804220e+18  9.804220e+18  ...     False     False   
296019  9.684320e+18  9.684320e+18  9.684320e+18  ...     False     False   
296020  5.119390e+18  5.119390e+18  5.119390e+18  ...     False     False   

       cooc_145  cooc_146  cooc_147 cooc_148  cooc_149  cooc_150  cooc_151  \
0         False     False     False    False     False     False     False   
1         False     False     False    False     False     False     False   
2         False     False     False    False     False     False     False   
3         False     False     False    False     False     False     False   
4         False     False     False    False     False     False     False   
...         ...       ...       ...      ...       ...       ...       ...   
296016    False     False     False    False     False     False     False   
296017    False     False     False    False     False     False     False   
296018    False     False     False    False     False     False     False   
296019    False     False     False    False     False     False     False   
296020    False     False     False    False     False     False     False   

        class  
0          16  
1         133  
2          16  
3          13  
4         133  
...       ...  
296016     19  
296017     13  
296018     16  
296019     19  
296020     96  

[296021 rows x 208 columns]

In [3]:
def city (series):
    if series == "Adelaide":
        return 0
    if series == "Amman":
        return 1 
    if series == "Amsterdam":
        return 2 
    if series == "Athens":
        return 3
    if series == "Auckland":
        return 4 
    if series == "Bahai":
        return 5 
    if series == "Bangkok":
        return 6
    if series == "Belem":
        return 7
    if series == "Berlin":
        return 8
    if series == "Bogota":
        return 9 
    if series == "Boise":
        return 10
    if series == "Bratislava":
        return 11 
    if series == "Bribane":
        return 12 
    if series == "Brunie":
        return 13
    if series == "Bussels":
        return 14
    if series == "Bucharest":
        return 15
    if series == "Buenos_Aires":
        return 16
    if series == "Cairo":
        return 17 
    if series == "Casablanca":
        return 18
    if series == "Chicago":
        return 19
    if series == "Copenhagen":
        return 20
    if series == "Cordoba":
        return 21
    if series == "Costa_Rica":
        return 22 
    if series == "Damascus":
        return 23
    if series == "Denver":
        return 24 
    if series == "Detroit":
        return 25 
    if series == "Dubia":
        return 26
    if series == "Dublin":
        return 27 
    if series == "Edmonton":
        return 28
    if series == "Fortaleza":
        return 29 
    if series == "Guam":
        return 30 
    if series == "Guayaquil":
        return 31
    if series == "Guyana":
        return 32
    if series == "Halifax":
        return 33 
    if series == "Helsinki":
        return 34 
    if series == "Ho_Chi_Minh":
        return 35
    if series == "Hobart":
        return 36
    if series == "Hong_Kong":
        return 37
    if series == "Honolulu":
        return 38
    if series == "Indianapolis":
        return 39
    if series == "Isle_of_Man":
        return 40
    if series == "Istanbul":
        return 41
    if series == "Jakarta":
        return 42
    if series == "Jerusalem":
        return 43
    if series == "Karachi":
        return 44
    if series == "Kiev":
        return 45 
    if series == "Kolkata":
        return 46
    if series == "Kuala_Lumpur":
        return 47
    if series == "Kuching":
        return 48
    if series == "Lisbon":
        return 49
    if series == "Ljubljana":
        return 50
    if series == "London":
        return 51
    if series == "Los_Angeles":
        return 52
    if series == "Louisville":
        return 53
    if series == "Luanda":
        return 54
    if series == "Luxembourg":
        return 56
    if series == "Madrid":
        return 57
    if series == "Manila":
        return 58
    if series == "Melbourne":
        return 59
    if series == "Monrovia":
        return 60
    if series == "Mexico_City":
        return 61
    if series == "Monterrey":
        return 62
    if series == "Montreal":
        return 63
    if series == "Moscow":
        return 64
    if series == "New_York":
        return 65
    if series == "Olso":
        return 66
    if series == "Nicosia":
        return 67
    if series == "Narobi":
        return 68
    if series == "Noumea":
        return 69
    if series == "Paris":
        return 70
    if series == "Perth":
        return 71
    if series == "Phnom_Penh":
        return 72
    if series == "Phoenix":
        return 73
    if series == "Prague":
        return 74
    if series == "Puerto_Rico":
        return 75
    if series == "Regina":
        return 76
    if series == "Reykjavik":
        return 77
    if series == "Rome":
        return 78
    if series == "Santiago":
        return 79
    if series == "Sao_Paulo":
        return 80
    if series == "Sarajevo":
        return 81
    if series == "Singapore":
        return 82
    if series == "Stockholm":
        return 83 
    if series == "Sydney":
        return 84
    if series == "Tahiti":
        return 85
    if series == "Taipei":
        return 86
    if series == "Tokyo":
        return 87
    if series == "Toronto":
        return 88
    if series == "Tripoli":
        return 89
    if series == "Tunis":
        return 90
    if series == "Vancouver":
        return 91
    if series == "Vienna":
        return 92
    if series == "Vilnius":
        return 93
    if series == "Warsaw":
        return 94
    if series == "Winnipeg":
        return 95
    if series == "Zagreb":
        return 96
    if series == "Zurich":
        return 97

In [4]:
pokemon['cityR'] = pokemon['city'].apply(city)
pokemon

pokemonId   latitude   longitude    appearedLocalTime  \
0              16  20.525745  -97.460829  2016-09-08T03:57:45   
1             133  20.523695  -97.461167  2016-09-08T03:57:37   
2              16  38.903590  -77.199780  2016-09-08T03:57:25   
3              13  47.665903 -122.312561  2016-09-08T03:56:22   
4             133  47.666454 -122.311628  2016-09-08T03:56:08   
...           ...        ...         ...                  ...   
296016         19  39.119052  -76.633194  2016-09-02T21:50:11   
296017         13  47.142237    7.244752  2016-09-02T21:50:08   
296018         16  42.258532  -87.881248  2016-09-02T21:50:07   
296019         19  26.230929  -98.194772  2016-09-02T21:50:07   
296020         96  50.144272   14.101178  2016-09-02T21:50:06   

                                 _id    cellId_90m   cellId_180m  \
0       NTgxMDkzOTk4MTM5MjUwMjIzNw==  9.645140e+18  9.645140e+18   
1       OTQ1NDgzODc1MjM3NDEzMTI2MQ==  9.645140e+18  9.645140e+18   
2       NTQ0OTQ0NDA1Nzg2ODg3OTg2OQ==  9.923200e+18  9.923200e+18   
3       NTU2MTU1NDM4NzA2MDk1MDcxNw==  6.093390e+18  6.093390e+18   
4       MTY2ODg4MTAzMTczMDE0MTUwNTM=  6.093390e+18  6.093390e+18   
...                              ...           ...           ...   
296016  MzM1MzA0NjEyNTIxNTQ1OTY5Mw==  9.923680e+18  9.923680e+18   
296017  NjA0NDc3OTcxMzk2NjM5MzgyMQ==  5.156090e+18  5.156090e+18   
296018  OTIzODI4ODYyOTUyMjI4NzI0NQ==  9.804220e+18  9.804220e+18   
296019  MTcwMjEwNDQ3Mjk1MjIwOTM5NDk=  9.684320e+18  9.684320e+18   
296020  MTMzNzM3NzQ1MTM4MTcwMTE0Njk=  5.119390e+18  5.119390e+18   

         cellId_370m   cellId_730m  cellId_1460m  ...  cooc_144  cooc_145  \
0       9.645140e+18  9.645140e+18  9.645140e+18  ...     False     False   
1       9.645140e+18  9.645140e+18  9.645140e+18  ...     False     False   
2       9.923200e+18  9.923200e+18  9.923200e+18  ...     False     False   
3       6.093390e+18  6.093390e+18  6.093390e+18  ...     False     False   
4       6.093390e+18  6.093390e+18  6.093390e+18  ...     False     False   
...              ...           ...           ...  ...       ...       ...   
296016  9.923680e+18  9.923680e+18  9.923680e+18  ...     False     False   
296017  5.156090e+18  5.156090e+18  5.156090e+18  ...     False     False   
296018  9.804220e+18  9.804220e+18  9.804220e+18  ...     False     False   
296019  9.684320e+18  9.684320e+18  9.684320e+18  ...     False     False   
296020  5.119390e+18  5.119390e+18  5.119390e+18  ...     False     False   

       cooc_146  cooc_147  cooc_148 cooc_149  cooc_150  cooc_151  class  cityR  
0         False     False     False    False     False     False     16   61.0  
1         False     False     False    False     False     False    133   61.0  
2         False     False     False    False     False     False     16   65.0  
3         False     False     False    False     False     False     13   52.0  
4         False     False     False    False     False     False    133   52.0  
...         ...       ...       ...      ...       ...       ...    ...    ...  
296016    False     False     False    False     False     False     19   65.0  
296017    False     False     False    False     False     False     13   97.0  
296018    False     False     False    False     False     False     16   19.0  
296019    False     False     False    False     False     False     19   19.0  
296020    False     False     False    False     False     False     96   74.0  

[296021 rows x 209 columns]

In [5]:
x = pokemon[[ 'latitude', 'longitude', 'cellId_90m', 'cellId_180m', 'cellId_370m', 'cellId_730m', 'cellId_1460m',
             'cellId_2920m', 'cellId_5850m', 'class']]
y = pokemon['cityR']

In [6]:
pokemon.dropna(inplace=True)

In [10]:
pokemon.describe()

pokemonId       latitude      longitude    cellId_90m  \
count  287960.000000  287960.000000  287960.000000  2.879600e+05   
mean       45.180549      38.572430     -37.625544  7.138899e+18   
std        39.580159      16.608224      66.839769  2.580550e+18   
min         1.000000     -42.984504    -159.332033  4.317270e+16   
25%        16.000000      33.859305     -90.933962  5.147408e+18   
50%        21.000000      42.008291     -72.751844  6.090680e+18   
75%        67.000000      49.272921      12.060180  9.759080e+18   
max       149.000000      68.099612     175.616223  1.228090e+19   

        cellId_180m   cellId_370m   cellId_730m  cellId_1460m  cellId_2920m  \
count  2.879600e+05  2.879600e+05  2.879600e+05  2.879600e+05  2.879600e+05   
mean   7.138899e+18  7.138899e+18  7.138899e+18  7.138899e+18  7.138899e+18   
std    2.580550e+18  2.580550e+18  2.580550e+18  2.580550e+18  2.580550e+18   
min    4.317270e+16  4.317270e+16  4.317270e+16  4.317270e+16  4.317270e+16   
25%    5.147408e+18  5.147408e+18  5.147408e+18  5.147408e+18  5.147408e+18   
50%    6.090680e+18  6.090680e+18  6.090680e+18  6.090680e+18  6.090680e+18   
75%    9.759080e+18  9.759080e+18  9.759080e+18  9.759080e+18  9.759080e+18   
max    1.228090e+19  1.228090e+19  1.228090e+19  1.228090e+19  1.228090e+19   

       cellId_5850m  ...  sunriseMinute  sunriseMinutesSince  \
count  2.879600e+05  ...  287960.000000        287960.000000   
mean   7.138899e+18  ...      28.168704           484.822805   
std    2.580550e+18  ...      15.085304           409.198842   
min    4.317260e+16  ...       0.000000          -444.000000   
25%    5.147408e+18  ...      16.000000           245.000000   
50%    6.090680e+18  ...      26.000000           561.000000   
75%    9.759080e+18  ...      41.000000           823.000000   
max    1.228090e+19  ...      59.000000          1092.000000   

       sunsetMinutesMidnight     sunsetHour   sunsetMinute  \
count          287960.000000  287960.000000  287960.000000   
mean             1175.670166      19.034262      33.614464   
std                29.464457       0.496558      15.316763   
min              1043.000000      17.000000       0.000000   
25%              1164.000000      19.000000      21.000000   
50%              1180.000000      19.000000      37.000000   
75%              1191.000000      19.000000      45.000000   
max              1251.000000      20.000000      59.000000   

       sunsetMinutesBefore  population_density  gymDistanceKm          class  \
count        287960.000000       287960.000000  287960.000000  287960.000000   
mean            291.667752         1338.640264       1.391239      45.180549   
std             410.821338         1920.087036      11.173780      39.580159   
min            -375.000000            0.000000       0.000591       1.000000   
25%             -52.000000          160.222840       0.182128      16.000000   
50%             221.000000          612.916900       0.331439      21.000000   
75%             538.000000         1754.262100       0.624360      67.000000   
max            1229.000000        10000.000000    3859.424664     149.000000   

               cityR  
count  287960.000000  
mean       57.494044  
std        23.150860  
min         0.000000  
25%        51.000000  
50%        65.000000  
75%        74.000000  
max        97.000000  

[8 rows x 32 columns]

In [12]:
pokemon.values()

array([[16, 20.525745, -97.460829, ..., False, 16, 61.0],
       [133, 20.523695, -97.461167, ..., False, 133, 61.0],
       [16, 38.90359, -77.19978, ..., False, 16, 65.0],
       ...,
       [16, 42.258532, -87.881248, ..., False, 16, 19.0],
       [19, 26.230929, -98.194772, ..., False, 19, 19.0],
       [96, 50.144272, 14.101178, ..., False, 96, 74.0]], dtype=object)

Train Test Split 

In [7]:
x_train, x_test, y_test, y_train = train_test_split(x,y, test_size = .4, random_state=101)
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

(177612, 10) (118409,)
(118409, 10) (177612,)


x_train dataset the 177612 rows and 10 columns, x_test dataset the 118490 rows and 10 columns 

Create Linear Regression Model 

In [8]:
lm = LinearRegression()
lm.fit(x_train, y_train)

ValueError: Input y contains NaN.